lms/djangoapps/grades/tests/test_course_grade_factory.py

```diff
@@ -124,6 +124,35 @@ def test_read_zero(self, assume_zero_enabled, create_if_needed):
             else:
                 self.assertIsNone(course_grade)
 
+    def test_read_optimization(self):
+        grade_factory = CourseGradeFactory()
+        with patch('lms.djangoapps.grades.course_data.get_course_blocks') as mocked_course_blocks:
+            mocked_course_blocks.return_value = self.course_structure
+            with mock_get_score(1, 2):
+                grade_factory.update(self.request.user, self.course, force_update_subsections=True)
+                self.assertEquals(mocked_course_blocks.call_count, 1)
+
+        with patch('lms.djangoapps.grades.course_data.get_course_blocks') as mocked_course_blocks:
+            with patch('lms.djangoapps.grades.subsection_grade.get_score') as mocked_get_score:
+                course_grade = grade_factory.read(self.request.user, self.course)
+                self.assertEqual(course_grade.percent, 0.5)  # make sure it's not a zero-valued course grade
+                self.assertFalse(mocked_get_score.called)  # no calls to CSM/submissions tables
+                self.assertFalse(mocked_course_blocks.called)  # no user-specific transformer calculation
```

Comment:
**Note to reviewers:** This test verifies that `get_course_blocks` (user-specific block-structure transforms) and `get_score` (problem-score retrieval from CSM/Submissions) no longer get called when the Course grade is read.